suprb/optimizer/solution/saga2/base.py

```diff
@@ -0,0 +1,170 @@
+        self.crossover_rate_min = crossover_rate_min
+        self.crossover_rate_current_min = crossover_rate_min
+        self.crossover_rate_max = crossover_rate_max
+        self.mutation = mutation
+        self.crossover = crossover
+        self.selection = selection
+        self.elitist_ratio = elitist_ratio
+
+    def _optimize(self, X: np.ndarray, y: np.ndarray):
+        self.fit_population(X, y)
+
+        self.n_elitists_ = int(self.population_size * self.elitist_ratio)
+
+        fitness_variance = np.var([i.fitness_ for i in self.population_])
+        self.fitness_variance_min = fitness_variance
+        self.fitness_variance_max = fitness_variance
+        self.old_gen = self.population_
+        self.fitness_mean = np.mean([i.fitness_ for i in self.population_])
+        self.fitness_min = np.min([i.fitness_ for i in self.population_])
+        self.fitness_max = np.max([i.fitness_ for i in self.population_])
+
+        for iter_ in range(self.n_iter):
+
+            # Elitism
+            elitists = sorted(self.population_, key=lambda i: i.fitness_, reverse=True)[:self.n_elitists_]
+
+            # Selection
+            parents = self.selection(population=self.population_, n=self.population_size,
+                                     random_state=self.random_state_)
+            
+            # Note that this expression swallows the last element, if `population_size` is odd
+            parent_pairs = map(lambda *x: x, *([iter(parents)] * 2))
+
+            # Crossover
+            children = list(flatten([(self.crossover(A, B, self.crossover_rate_current_min, self.crossover_rate_max, self.fitness_mean, self.fitness_min, self.fitness_max, random_state=self.random_state_),
+                                      self.crossover(B, A, self.crossover_rate_current_min, self.crossover_rate_max, self.fitness_mean, self.fitness_min, self.fitness_max, random_state=self.random_state_))
+                                     for A, B in parent_pairs]))
+            # If `population_size` is odd, we add the solution not selected for reproduction directly
+            if self.population_size % 2 != 0:
+                children.append(parents[-1])
+
+            # Mutation
+            for child in children: child.fit(X, y)
+            mutated_children = [self.mutation(child, self.mutation_rate_min, self.mutation_rate_current_max, self.fitness_mean, self.fitness_max, random_state=self.random_state_) for child in children]
+
+            # Keep copy of old generation for comparison
+            self.old_gen = self.population_
+
+            # Replacement
+            self.population_ = elitists
+            self.population_.extend(mutated_children)
+
+            self.fit_population(X, y)
+            self.fitness_mean = np.mean(([i.fitness_ for i in self.population_]))
+            self.fitness_min = np.min(([i.fitness_ for i in self.population_]))
+            self.fitness_max = np.max(([i.fitness_ for i in self.population_]))
```

Comment:
`[i.fitness_ for i in self.population_] `save this in a variable at the top of the function, then we don't need to calculate it 3 times